airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/event/CatalogValidator.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.core.event
+
+import io.airbyte.cdk.core.config.AirbyteConfiguredCatalog
+import io.airbyte.cdk.core.context.env.ConnectorConfigurationPropertySource
+import io.airbyte.cdk.core.operation.OperationType
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Requires
+import io.micronaut.context.annotation.Value
+import io.micronaut.context.event.ApplicationEventListener
+import io.micronaut.context.event.StartupEvent
+import jakarta.inject.Singleton
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * Event listener that validates the Airbyte configured catalog, if present. This listener is
+ * executed on application start.
+ */
+@Singleton
+@Requires(bean = AirbyteConfiguredCatalog::class)
+@Requires(property = ConnectorConfigurationPropertySource.CONNECTOR_OPERATION)
+class CatalogValidator(
+    @Value("\${micronaut.application.name}") private val connectorName: String,
+    @Value("\${airbyte.connector.operation}") private val operation: String,
+    private val airbyteConfiguredCatalog: AirbyteConfiguredCatalog,
+) : ApplicationEventListener<StartupEvent> {
+    companion object {
+        var emptyCatalog: ConfiguredAirbyteCatalog = ConfiguredAirbyteCatalog()
+    }
+
+    override fun onApplicationEvent(event: StartupEvent) {
+        if (requiresCatalog()) {
+            if (emptyCatalog != airbyteConfiguredCatalog.getConfiguredCatalog()) {
+                logger.info { "$connectorName connector configured catalog is valid." }
+            } else {
+                throw IllegalArgumentException("Configured catalog is not valid.")
+            }
+        }
+    }
```

Comment:
This is in the same vein as other comments: we _expect_ the catalog to be well-formed, can we get rid of this CatalogValidator thing entirely and barf an exception earlier in the getter or wherever? My perhaps naive take is that feels like a lot of complexity to guard against something that's infrequent and unexpected. I think it's fine to fail hard in a slightly ugly way.

This comment also applies elsewhere.

Reply:
@postamar I'm not sure if we can hook into the configuration property mapping done by Micronaut.  We might be able to add a `@PostConstruct` to the catalog class that is populated by Micronaut and do the validation there instead of having these handlers.   Note that I do not like this either -- this is merely a port of the logic from `IntegrationRunner` that does validation on argument passed to the command.

Reply:
Sounds like more boilerplate, when we want less. Never mind, we'll live with this.